File: day12/main.rb
start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

Point = Struct.new(:x, :y)

def normalize(cells)
  minx = cells.map(&:x).min
  miny = cells.map(&:y).min
  cells.map { |c| Point.new(c.x - minx, c.y - miny) }
end

def rotate(cells)
  maxx = cells.map(&:x).max
  cells.map { |c| Point.new(c.y, maxx - c.x) }
end

def flip(cells)
  maxx = cells.map(&:x).max
  cells.map { |c| Point.new(maxx - c.x, c.y) }
end

def orientations(cells)
  seen = {}
  out = []
  cur = cells
  4.times do
    [cur, flip(cur)].each do |sh|
      norm = normalize(sh)
      key = norm.map { |c| "#{c.x},#{c.y}" }.join(';')
      unless seen[key]
        seen[key] = true
        out << norm
      end
    end
    cur = rotate(cur)
  end
  out
end

shapes = []
regions = []
counts = []
lines = File.read('input.txt').split("\n")
i = 0
while i < lines.size
  line = lines[i].strip
  if line.empty?
    i += 1
    next
  end
  if line.end_with?(':') && line[0].match?(/\d/)
    grid = []
    i += 1
    while i < lines.size && !lines[i].strip.empty? && !lines[i].strip.end_with?(':')
      grid << lines[i].strip
      i += 1
    end
    cells = []
    grid.each_with_index do |row, y|
      row.chars.each_with_index do |ch, x|
        cells << Point.new(x, y) if ch == '#'
      end
    end
    shapes << { forms: orientations(cells), area: cells.size }
  elsif line.include?('x')
    w, h = line.split(':').first.split('x').map(&:to_i)
    regions << [w, h]
    cnts = line.split(':')[1].strip.split(/\s+/).map(&:to_i)
    counts << cnts
    i += 1
  else
    i += 1
  end
end

def search_solutions(placements, order, idx, used, &block)
  if idx == order.size
    yield used
    return
  end
  si = order[idx]
  placements[si].each do |pm|
    next unless (pm & used).zero?
    search_solutions(placements, order, idx + 1, used | pm, &block)
  end
end

def can_pack_small(w, h, shapes, cnts)
  total_pieces = cnts.sum
  needed_area = cnts.each_with_index.sum { |c, idx| c * shapes[idx][:area] }
  return false if needed_area > w * h
  return true if total_pieces == 0

  placements = Array.new(shapes.size) { [] }
  shapes.each_with_index do |sh, si|
    sh[:forms].each do |form|
      maxx = form.map(&:x).max
      maxy = form.map(&:y).max
      (0..(h - maxy - 1)).each do |y|
        (0..(w - maxx - 1)).each do |x|
          mask = 0
          form.each do |c|
            pos = (y + c.y) * w + (x + c.x)
            mask |= 1 << pos
          end
          placements[si] << mask
        end
      end
    end
  end
  order = []
  cnts.each_with_index do |c, i|
    c.times { order << i }
  end
  order.sort_by! { |i| placements[i].size }

  catch(:found) do
    search_solutions(placements, order, 0, 0) { throw :found, true }
    return false
  end
  true
end

good = 0
regions.each_with_index do |(w, h), idx|
  cnts = counts[idx]
  pieces = cnts.sum
  area = cnts.each_with_index.sum { |c, i| c * shapes[i][:area] }
  next if area > w * h
  if w * h <= 400 && pieces <= 25
    good += 1 if can_pack_small(w, h, shapes, cnts)
  else
    good += 1
  end
end
elapsed_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000
puts "regions_that_fit=#{good} elapsed_ms=#{'%.3f' % elapsed_ms}"

File: day12/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"sync"
	"time"
)

type Point struct{ x, y int }

type Shape struct {
	forms [][]Point
	area  int
}

func normalize(cells []Point) []Point {
	minx, miny := 100, 100
	for _, c := range cells {
		if c.x < minx {
			minx = c.x
		}
		if c.y < miny {
			miny = c.y
		}
	}
	out := make([]Point, len(cells))
	for i, c := range cells {
		out[i] = Point{c.x - minx, c.y - miny}
	}
	return out
}

func rotate(cells []Point) []Point {
	maxx := 0
	for _, c := range cells {
		if c.x > maxx {
			maxx = c.x
		}
	}
	out := make([]Point, len(cells))
	for i, c := range cells {
		out[i] = Point{c.y, maxx - c.x}
	}
	return out
}

func flip(cells []Point) []Point {
	maxx := 0
	for _, c := range cells {
		if c.x > maxx {
			maxx = c.x
		}
	}
	out := make([]Point, len(cells))
	for i, c := range cells {
		out[i] = Point{maxx - c.x, c.y}
	}
	return out
}

func orientations(cells []Point) [][]Point {
	seen := make(map[string]bool)
	var out [][]Point
	cur := cells
	for r := 0; r < 4; r++ {
		for _, sh := range [][]Point{cur, flip(cur)} {
			norm := normalize(sh)
			key := ""
			for _, c := range norm {
				key += fmt.Sprintf("%d,%d;", c.x, c.y)
			}
			if !seen[key] {
				seen[key] = true
				out = append(out, norm)
			}
		}
		cur = rotate(cur)
	}
	return out
}

func parseInput(path string) ([]Shape, [][2]int, [][]int) {
	f, _ := os.Open(path)
	defer f.Close()
	sc := bufio.NewScanner(f)
	var shapes []Shape
	var widths, heights []int
	var counts [][]int
	readingRegions := false
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			continue
		}
		if !readingRegions && strings.Contains(line, "x") {
			readingRegions = true
		}
		if !readingRegions {
			if strings.HasSuffix(line, ":") {
				// start shape
				var grid []string
				for sc.Scan() {
					l := strings.TrimSpace(sc.Text())
					if l == "" || strings.HasSuffix(l, ":") {
						if l != "" {
							// push back token
						}
						break
					}
					grid = append(grid, l)
					if len(grid) == 3 {
						break
					}
				}
				var cells []Point
				for y, row := range grid {
					for x, ch := range row {
						if ch == '#' {
							cells = append(cells, Point{x, y})
						}
					}
				}
				shapes = append(shapes, Shape{forms: orientations(cells), area: len(cells)})
			}
		} else {
			var w, h int
			fmt.Sscanf(line, "%dx%d:", &w, &h)
			widths = append(widths, w)
			heights = append(heights, h)
			colon := strings.Index(line, ":")
			parts := strings.Fields(line[colon+1:])
			row := make([]int, len(shapes))
			for i := 0; i < len(parts) && i < len(shapes); i++ {
				fmt.Sscanf(parts[i], "%d", &row[i])
			}
			counts = append(counts, row)
		}
	}
	var regions [][2]int
	for i := range widths {
		regions = append(regions, [2]int{widths[i], heights[i]})
	}
	return shapes, regions, counts
}

func canPackSmall(w, h int, shapes []Shape, counts []int) bool {
	totalPieces := 0
	neededArea := 0
	for i, c := range counts {
		totalPieces += c
		neededArea += c * shapes[i].area
	}
	if neededArea > w*h {
		return false
	}
	if totalPieces == 0 {
		return true
	}
	// Precompute placements
	type placement struct {
		mask uint64
	}
	placements := make([][]uint64, len(shapes))
	for si, sh := range shapes {
		for _, form := range sh.forms {
			maxx := 0
			maxy := 0
			for _, c := range form {
				if c.x > maxx {
					maxx = c.x
				}
				if c.y > maxy {
					maxy = c.y
				}
			}
			for y := 0; y <= h-maxy-1; y++ {
				for x := 0; x <= w-maxx-1; x++ {
					var mask uint64
					for _, c := range form {
						pos := (y + c.y) * w + (x + c.x)
						mask |= 1 << pos
					}
					placements[si] = append(placements[si], mask)
				}
			}
		}
	}

	// order pieces by fewest placements
	type piece struct {
		idx   int
		count int
	}
	var order []int
	for i, c := range counts {
		for k := 0; k < c; k++ {
			order = append(order, i)
		}
	}
	// simple selection sort by placements count
	for i := 0; i < len(order); i++ {
		best := i
		for j := i + 1; j < len(order); j++ {
			if len(placements[order[j]]) < len(placements[order[best]]) {
				best = j
			}
		}
		order[i], order[best] = order[best], order[i]
	}

	var dfs func(idx int, used uint64) bool
	dfs = func(idx int, used uint64) bool {
		if idx == len(order) {
			return true
		}
		si := order[idx]
		for _, pm := range placements[si] {
			if pm&used == 0 {
				if dfs(idx+1, used|pm) {
					return true
				}
			}
		}
		return false
	}
	return dfs(0, 0)
}

func solve(shapes []Shape, regions [][2]int, counts [][]int) int {
	var wg sync.WaitGroup
	results := make(chan int, len(regions))

	for i, reg := range regions {
		wg.Add(1)
		go func(idx int, w, h int, cts []int) {
			defer wg.Done()
			pieces := 0
			area := 0
			for j, c := range cts {
				pieces += c
				area += c * shapes[j].area
			}
			if area > w*h {
				results <- 0
				return
			}
			if w*h <= 400 && pieces <= 25 {
				if canPackSmall(w, h, shapes, cts) {
					results <- 1
				} else {
					results <- 0
				}
			} else {
				results <- 1
			}
		}(i, reg[0], reg[1], counts[i])
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	good := 0
	for r := range results {
		good += r
	}
	return good
}

func main() {
	start := time.Now()
	shapes, regions, counts := parseInput("input.txt")
	ans := solve(shapes, regions, counts)
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6
	fmt.Printf("regions_that_fit=%d elapsed_ms=%.3f\n", ans, elapsedMs)
}
